lib/srv/uacc/uacc_linux.go

```diff
@@ -0,0 +1,144 @@
+var accountDb sync.Mutex
+
+// Max length of username and hostname as defined by glibc.
+const nameMaxLen = 255
+
+// Open writes a new entry to the utmp database with a tag of `USER_PROCESS`.
+// This should be called when an interactive session is started.
+//
+// `username`: Name of the user the interactive session is running under.
+// `hostname`: Name of the system the user is logged into.
+// `remoteAddrV6`: IPv6 address of the remote host.
+// `ttyName`: Name of the TTY without the `/dev/` prefix.
+func Open(username, hostname string, remote net.IP, ttyName string) error {
+	rawV6 := remote.To16()
+	groupedV6 := [4]int32{}
+	for i := range groupedV6 {
+		groupedV6[i] = int32(binary.LittleEndian.Uint32(rawV6[i*4 : (i+1)*4]))
+	}
+
+	// String parameter validation.
+	if len(username) > nameMaxLen {
+		return trace.BadParameter("username length exceeds OS limits")
+	}
+	if len(hostname) > nameMaxLen {
+		return trace.BadParameter("hostname length exceeds OS limits")
+	}
+	if len(ttyName) > (int)(C.max_len_tty_name()-1) {
+		return trace.BadParameter("tty name length exceeds OS limits")
+	}
+
+	// Convert Go strings into C strings that we can pass over ffi.
+	cUsername := C.CString(username)
+	defer C.free(unsafe.Pointer(cUsername))
+	cHostname := C.CString(hostname)
+	defer C.free(unsafe.Pointer(cHostname))
+	cTtyName := C.CString(ttyName[5:])
+	defer C.free(unsafe.Pointer(cTtyName))
+	cIDName := C.CString(ttyName[9:])
+	defer C.free(unsafe.Pointer(cIDName))
+
+	// Convert IPv6 array into C integer format.
+	CInts := [4]C.int{}
+	for i := 0; i < 4; i++ {
+		CInts[i] = (C.int)(groupedV6[i])
+	}
+
+	timestamp := time.Now()
+	secondsElapsed := (C.int32_t)(timestamp.Unix())
+	microsFraction := (C.int32_t)((timestamp.UnixNano() % int64(time.Second)) / int64(time.Microsecond))
+
+	accountDb.Lock()
+	status := C.uacc_add_utmp_entry(cUsername, cHostname, &CInts[0], cTtyName, cIDName, secondsElapsed, microsFraction)
+	accountDb.Unlock()
+
+	switch status {
+	case C.UACC_UTMP_MISSING_PERMISSIONS:
+		return trace.Errorf("InteractiveSessionOpened missing permissions to write to utmp/wtmp")
+	case C.UACC_UTMP_WRITE_ERROR:
+		return trace.Errorf("InteractiveSessionOpened failed to add entry to utmp database")
+	case C.UACC_UTMP_FAILED_OPEN:
+		return trace.Errorf("InteractiveSessionOpened failed to open user account database")
+	default:
+		if status != 0 {
+			return trace.Errorf("unknown error with code %d", status)
+		}
+
+		return nil
+	}
+}
+
+// Close marks an entry in the utmp database as DEAD_PROCESS.
+// This should be called when an interactive session exits.
+//
+// The `ttyName` parameter must be the name of the TTY including the `/dev/` prefix.
+func Close(ttyName string) error {
+	// String parameter validation.
+	if len(ttyName) > (int)(C.max_len_tty_name()-1) {
+		return trace.BadParameter("tty name length exceeds OS limits")
+	}
+
+	// Convert Go strings into C strings that we can pass over ffi.
+	cTTYName := C.CString(ttyName)
```

Comment:
should the `/dev/` prefix be stripped here too?

Reply:
Good catch, yes it should. Glibc might do this implicitly but it doesn't look like that is standardized so I'll do it here too.